paddle/gserver/dataproviders/PyDataProvider2.cpp

```diff
@@ -496,8 +507,8 @@ class PyDataProvider2 : public DataProvider {
    * Resetting the PyDataProvider. May start reading thread here.
    */
   virtual void reset() {
-    DataProvider::reset();
     resetImpl(true);
+    DataProvider::reset();
```

Comment:
DataProvider::reset() should be invoked at the end of reset(). Because the DataProvider::reset() will invoke getNextBatchInternal in other thread right now.